File: app/src/main/java/io/builderscon/conference2017/view/SessionsFragment.java
package io.builderscon.conference2017.view;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.graphics.Point;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.SystemClock;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.support.v4.app.Fragment;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v7.widget.RecyclerView;
import android.view.Display;
import android.view.GestureDetector;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.lucasr.twowayview.TwoWayLayoutManager;
import org.lucasr.twowayview.widget.DividerItemDecoration;
import org.lucasr.twowayview.widget.SpannableGridLayoutManager;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;
import java.util.List;
import java.util.Map;

import io.builderscon.client.model.Room;
import io.builderscon.conference2017.R;
import io.builderscon.conference2017.databinding.FragmentSessionsBinding;
import io.builderscon.conference2017.databinding.ViewSessionCellBinding;
import io.builderscon.conference2017.view.customview.TouchlessTwoWayView;


public class SessionsFragment extends Fragment {

    SessionsViewModel viewModel = new SessionsViewModel();

    private FragmentSessionsBinding binding;

    private SessionsAdapter adapter;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        binding = FragmentSessionsBinding.inflate(inflater, container, false);
        binding.setViewModel(viewModel);

        initView();
        return binding.getRoot();
    }

    @Override
    public void onResume() {
        super.onResume();
        showSessions();
    }

    private int getScreenWidth() {
        Display display = getActivity().getWindowManager().getDefaultDisplay();
        Point size = new Point();
        display.getSize(size);
        return size.x;
    }

    private void showSessions() {
        this.renderSessions(viewModel.getSessions(getActivity(), getArguments().getInt("tabIndex")));
    }

    private void initView() {

        binding.recyclerView.setHasFixedSize(true);

        final Drawable divider = ResourcesCompat.getDrawable(getResources(), R.drawable.divider, null);
        binding.recyclerView.addItemDecoration(new DividerItemDecoration(divider));

        adapter = new SessionsAdapter(getActivity());
        binding.recyclerView.setAdapter(adapter);

        final ClickGestureCanceller clickCanceller = new ClickGestureCanceller(getActivity(), binding.recyclerView);

        binding.root.setOnTouchListener((v, event) -> {
            clickCanceller.sendCancelIfScrolling(event);

            MotionEvent e = MotionEvent.obtain(event);
            e.setLocation(e.getX() + binding.root.getScrollX(), e.getY() - binding.headerRow.getHeight());
            binding.recyclerView.forceToDispatchTouchEvent(e);
            return false;
        });

        ViewUtil.addOneTimeOnGlobalLayoutListener(binding.headerRow, () -> {
            if (binding.headerRow.getHeight() > 0) {
                binding.recyclerView.getLayoutParams().height = binding.root.getHeight() - binding.border.getHeight()
                        - binding.headerRow.getHeight();
                binding.recyclerView.requestLayout();
                return true;
            } else {
                return false;
            }
        });
    }

    private void renderSessions(List<SessionViewModel> adjustedSessionViewModels) {
        List<Date> stimes = viewModel.getStimes();
        List<Room> rooms = viewModel.getRooms();
        Map<String, String> tracks = viewModel.getTracksMap();

        int sessionsTableWidth = getScreenWidth();
        int minWidth = (int) getResources().getDimension(R.dimen.session_table_min_width);
        if (rooms.size() > 2 && sessionsTableWidth < minWidth) {
            sessionsTableWidth = minWidth;
        }
        binding.recyclerView.setMinimumWidth(sessionsTableWidth);

        RecyclerView.LayoutManager lm = new SpannableGridLayoutManager(
                TwoWayLayoutManager.Orientation.VERTICAL, rooms.size(), stimes.size());
        binding.recyclerView.setLayoutManager(lm);

        renderHeaderRow(rooms, tracks);
        adapter.reset(adjustedSessionViewModels);
    }

    private void renderHeaderRow(List<Room> rooms, Map<String, String> tracks) {
        if (binding.headerRow.getChildCount() == 0) {
            for (Room room : rooms) {
                View view = LayoutInflater.from(getActivity()).inflate(R.layout.view_sessions_header_cell, null);
                TextView txtRoomName = view.findViewById(R.id.txt_room_name);
                txtRoomName.setText(tracks.get(room.getId()));
                LinearLayout.LayoutParams params =
                        new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.MATCH_PARENT, 1f);
                txtRoomName.setLayoutParams(params);
                binding.headerRow.addView(view);
            }
        }
    }

    private static class ClickGestureCanceller {

        private GestureDetector gestureDetector;

        ClickGestureCanceller(final Context context, final TouchlessTwoWayView targetView) {
            gestureDetector = new GestureDetector(context, new GestureDetector.OnGestureListener() {
                private boolean ignoreMotionEventOnScroll = false;

                @Override
                public boolean onDown(MotionEvent motionEvent) {
                    ignoreMotionEventOnScroll = true;
                    return false;
                }

                @Override
                public boolean onScroll(MotionEvent motionEvent, MotionEvent motionEvent1, float v, float v1) {

                    // Send cancel event for item clicked when horizontal scrolling.
                    if (ignoreMotionEventOnScroll) {
                        final long now = SystemClock.uptimeMillis();
                        MotionEvent cancelEvent = MotionEvent.obtain(now, now,
                                MotionEvent.ACTION_CANCEL, 0.0f, 0.0f, 0);
                        targetView.forceToDispatchTouchEvent(cancelEvent);
                        ignoreMotionEventOnScroll = false;
                    }

                    return false;
                }

                @Override
                public void onShowPress(MotionEvent motionEvent) {
                    // Do nothing
                }

                @Override
                public boolean onSingleTapUp(MotionEvent motionEvent) {
                    // Do nothing
                    return false;
                }

                @Override
                public void onLongPress(MotionEvent motionEvent) {
                }

                @Override
                public boolean onFling(MotionEvent motionEvent, MotionEvent motionEvent1, float v, float v1) {
                    return false;
                }
            });
        }

        void sendCancelIfScrolling(MotionEvent event) {
            gestureDetector.onTouchEvent(event);
        }
    }

    public class SessionsAdapter extends ArrayRecyclerAdapter<SessionViewModel, BindingHolder<ViewSessionCellBinding>> {

        SessionsAdapter(@NonNull Context context) {
            super(context);
        }

        @Override
        public BindingHolder<ViewSessionCellBinding> onCreateViewHolder(ViewGroup parent, int viewType) {
            return new BindingHolder<>(getContext(), parent, R.layout.view_session_cell);
        }

        @Override
        public void onBindViewHolder(BindingHolder<ViewSessionCellBinding> holder, int position) {
            SessionViewModel viewModel = getItem(position);
            holder.binding.setViewModel(viewModel);
            holder.binding.executePendingBindings();
        }
    }

}

class BindingHolder<T extends ViewDataBinding> extends RecyclerView.ViewHolder {

    final T binding;

    BindingHolder(@NonNull Context context, @NonNull ViewGroup parent, @LayoutRes int layoutResId) {
        super(LayoutInflater.from(context).inflate(layoutResId, parent, false));
        binding = DataBindingUtil.bind(itemView);
    }
}


abstract class ArrayRecyclerAdapter<T, VH extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<VH> {

    private final List<T> list;

    final Context context;

    ArrayRecyclerAdapter(@NonNull Context context) {
        this(context, new ArrayList<>());
    }


    private ArrayRecyclerAdapter(@NonNull Context context, @NonNull List<T> list) {
        this.context = context;
        this.list = list;
    }

    @UiThread
    void reset(Collection<T> items) {
        clear();
        addAll(items);
        notifyDataSetChanged();
    }

    private void clear() {
        list.clear();
    }

    private void addAll(Collection<T> items) {
        list.addAll(items);
    }

    T getItem(int position) {
        return list.get(position);
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public Context getContext() {
        return context;
    }

}

File: app/src/main/kotlin/io/builderscon/conference2017/view/fragment/SessionsFragment.kt
package io.builderscon.conference2017.view.fragment

import android.content.Context
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.os.SystemClock
import android.support.annotation.LayoutRes
import android.support.annotation.UiThread
import android.support.v4.app.Fragment
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.widget.RecyclerView
import android.view.*
import android.widget.LinearLayout
import android.widget.TextView
import io.builderscon.client.model.Room
import io.builderscon.conference2017.R
import io.builderscon.conference2017.databinding.FragmentSessionsBinding
import io.builderscon.conference2017.databinding.ViewSessionCellBinding
import io.builderscon.conference2017.extension.getScreenWidth
import io.builderscon.conference2017.view.customview.TouchlessTwoWayView
import io.builderscon.conference2017.viewmodel.SessionViewModel
import io.builderscon.conference2017.viewmodel.SessionsViewModel
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch
import org.lucasr.twowayview.TwoWayLayoutManager
import org.lucasr.twowayview.widget.DividerItemDecoration
import org.lucasr.twowayview.widget.SpannableGridLayoutManager
import java.util.*


class SessionsFragment : Fragment() {
    internal var viewModel = SessionsViewModel()

    lateinit var binding: FragmentSessionsBinding

    lateinit var adapter: SessionsFragment.SessionsAdapter

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        setHasOptionsMenu(true)
        binding = FragmentSessionsBinding.inflate(inflater!!, container, false)
        binding.viewModel = viewModel

        initView()
        return binding.getRoot()
    }

    override fun onResume() {
        super.onResume()
        showSessions()
    }

    private fun showSessions() {
        launch(UI) {
            val sessions = async(CommonPool) {
                viewModel.getSessions(arguments.getInt("tabIndex"))
            }.await()
            renderSessions(sessions)
        }
    }

    private fun initView() {

        binding.recyclerView.setHasFixedSize(true)

        val divider = ResourcesCompat.getDrawable(resources, R.drawable.divider, null)
        binding.recyclerView.addItemDecoration(DividerItemDecoration(divider))

        adapter = SessionsAdapter(activity)
        binding.recyclerView.adapter = adapter

        val clickCanceller = ClickGestureCanceller(activity, binding.recyclerView)

        binding.root.setOnTouchListener { _, event ->
            clickCanceller.sendCancelIfScrolling(event)

            val e = MotionEvent.obtain(event)
            e.setLocation(e.x + binding.root.scrollX, e.y - binding.headerRow.height)
            binding.recyclerView.forceToDispatchTouchEvent(e)
            false
        }

    }

    private fun renderSessions(adjustedSessionViewModels: List<SessionViewModel>) {
        val sTimes = viewModel.sTimes
        val rooms = viewModel.rooms
        val tracks = viewModel.tracksMap

        var sessionsTableWidth = activity.getScreenWidth()
        val minWidth = resources.getDimension(R.dimen.session_table_min_width).toInt()
        if (rooms.size > 2 && sessionsTableWidth < minWidth) {
            sessionsTableWidth = minWidth
        }
        binding.recyclerView.minimumWidth = sessionsTableWidth

        val lm = SpannableGridLayoutManager(
                TwoWayLayoutManager.Orientation.VERTICAL, rooms.size, sTimes.size)
        binding.recyclerView.layoutManager = lm

        renderHeaderRow(rooms, tracks)
        adapter.reset(adjustedSessionViewModels)
    }

    private fun renderHeaderRow(rooms: List<Room>, tracks: Map<String, String>) {
        if (binding.headerRow.childCount == 0) {
            for ((id) in rooms) {
                val view = LayoutInflater.from(activity).inflate(R.layout.view_sessions_header_cell, null)
                val txtRoomName = view.findViewById<TextView>(R.id.txt_room_name)
                txtRoomName.text = tracks[id]?.replace("(", "\n(")
                val params = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.MATCH_PARENT, 1f)
                txtRoomName.layoutParams = params
                binding.headerRow.addView(view)
            }
        }
    }

    private class ClickGestureCanceller internal constructor(context: Context, targetView: TouchlessTwoWayView) {

        private val gestureDetector: GestureDetector

        init {
            gestureDetector = GestureDetector(context, object : GestureDetector.OnGestureListener {
                private var ignoreMotionEventOnScroll = false

                override fun onDown(motionEvent: MotionEvent): Boolean {
                    ignoreMotionEventOnScroll = true
                    return false
                }

                override fun onScroll(motionEvent: MotionEvent, motionEvent1: MotionEvent, v: Float, v1: Float): Boolean {

                    // Send cancel event for item clicked when horizontal scrolling.
                    if (ignoreMotionEventOnScroll) {
                        val now = SystemClock.uptimeMillis()
                        val cancelEvent = MotionEvent.obtain(now, now,
                                MotionEvent.ACTION_CANCEL, 0.0f, 0.0f, 0)
                        targetView.forceToDispatchTouchEvent(cancelEvent)
                        ignoreMotionEventOnScroll = false
                    }

                    return false
                }

                override fun onShowPress(motionEvent: MotionEvent) {
                    // Do nothing
                }

                override fun onSingleTapUp(motionEvent: MotionEvent): Boolean {
                    // Do nothing
                    return false
                }

                override fun onLongPress(motionEvent: MotionEvent) {}

                override fun onFling(motionEvent: MotionEvent, motionEvent1: MotionEvent, v: Float, v1: Float): Boolean {
                    return false
                }
            })
        }

        internal fun sendCancelIfScrolling(event: MotionEvent) {
            gestureDetector.onTouchEvent(event)
        }
    }

    inner class SessionsAdapter internal constructor(context: Context) : ArrayRecyclerAdapter<SessionViewModel, BindingHolder<ViewSessionCellBinding>>(context) {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BindingHolder<ViewSessionCellBinding> {
            return BindingHolder(context, parent, R.layout.view_session_cell)
        }

        override fun onBindViewHolder(holder: BindingHolder<ViewSessionCellBinding>, position: Int) {
            val viewModel = getItem(position)
            holder.binding.viewModel = viewModel
            holder.binding.executePendingBindings()
        }
    }


}

class BindingHolder<out T : ViewDataBinding>(context: Context, parent: ViewGroup, @LayoutRes layoutResId: Int) : RecyclerView.ViewHolder(LayoutInflater.from(context).inflate(layoutResId, parent, false)) {
    val binding: T = DataBindingUtil.bind<T>(itemView)
}


abstract class ArrayRecyclerAdapter<T, VH : RecyclerView.ViewHolder> private constructor(val context: Context, private val list: MutableList<T>) : RecyclerView.Adapter<VH>() {

    constructor(context: Context) : this(context, ArrayList<T>()) {}

    @UiThread
    fun reset(items: Collection<T>) {
        clear()
        addAll(items)
        notifyDataSetChanged()
    }

    private fun clear() {
        list.clear()
    }

    private fun addAll(items: Collection<T>) {
        list.addAll(items)
    }

    fun getItem(position: Int): T {
        return list[position]
    }

    override fun getItemCount(): Int {
        return list.size
    }

}

